### ISSUE TITLE
encoding of CONNECTION_CLOSE reason phrases

### ISSUE THREAD

**Issue by marten-seemann** [quicwg/base-drafts#1990]
There recently was a Human Rights review of the QUIC protocol, which identified a potential issue with the encoding of CONNECTION_CLOSE frames: https://tools.ietf.org/id/draft-martini-hrpc-quichr-00.html#internationalization.

I don't know how many people read this document, so I'm opening this issue. Not sure if we actually need to do something about it, but we should at least consider their recommendation.

**Comment by mikkelfj** [comment]
I think MUST UTF-8 is fine, this also solves endian issues.

The language SHOULD br english when interoperability is required, but MAY be any language that all endpoints are willing to accept within a closed deployment.

Specifying language codes isn’t really helpful as you can’t really make use of the text if you don’t recognize it immediately, and it just complicates matters.

**Comment by martinthomson** [comment]
The problem here is that tagging isn't that useful unless you also couple it with negotiation.  We can annotate the string with a language tag, but that doesn't make the text more comprehensible if the chosen language isn't understood.  If the endpoint is able to generate translated strings, then it doesn't know which one to use.  It's possible that we could signal the preferred language of endpoints in transport parameters, but that leaves clients exposing their language preferences in the clear.  

Then we need to consider that the language preferences of the endpoint might not even be relevant.  This information is largely only useful during debugging, so it is the language preference of the stack developer that is likely to matter.  That said, if the messages were passed to the stack developer, that is likely to represent a risk to the privacy of the endpoints or those using those endpoints.  And there is a risk of misinterpretation if the language preference of the person using the endpoint differs from that of the stack developer.

In other words, this isn't a trivial problem.  We could just tag the message and leave it to online translation tools and the like to sort things out, I guess.  But @mikkelfj's conclusion is the one we took in HTTP/2, though we chose to say nothing in that case.

**Comment by kazuho** [comment]
What @martinthomson says.

The field is purely for debugging, and we do not need to internationalize it.

I am fine with changing the definition from "human-readable explanation that SHOULD be UTF-8" to "a series of octets that can be anything" if doing so removes the misconception that the information contained here requires I18N.

**Comment by marten-seemann** [comment]
I agree that this is purely for debugging, and should probably never be displayed to the user, so I18n shouldn't be an issue.

Is there any reason not to require this to be UTF-8 though? I think it would help debugging interoperability testing (and error logging) if you can rely on the charset.

**Comment by kazuho** [comment]
> Is there any reason not to require this to be UTF-8 though? I think it would help debugging interoperability testing (and error logging) if you can rely on the charset.

If we say "MUST be UTF-8", then the natural outcome would be that people would expect QUIC stacks to validate if the reason phase contains a valid UTF-8 string. I do not think we should have such a requirement.

Current text saying "SHOULD be UTF-8" is fine in this respect because QUIC stacks can simply emit the binary octets that it has received to the application.

Note that most (if not all) of HTTP logging does not deal with charsets; for example a URL can contain any charset, and they are logged as bytes.

**Comment by martinthomson** [comment]
You can't realistically generate an error if CONNECTION_CLOSE is bad, so I don't have a problem saying that it contains a UTF-8 string.

**Comment by mikkelfj** [comment]
If it's UTF-8 it is much simpler to dump in error logs, even if you don't bother validate it. Some logs tend to use JSON which is practically always UTF-8. If you need binary, just base64 encode it.

**Comment by mikkelfj** [comment]
A binary format is a buffer overrun attack waiting to happen.

**Comment by janaiyengar** [comment]
I am inclined to leave the text as it is. The primary motivation for i18n
is for user-visible text, and this text is not really it. For devs, I can't
imagine that language tags will help when the entire protocol is in
English. I'm fine with a SHOULD for UTF-8, and leave the implications of
non-UTF-8 encodings to implementations.

**Comment by kazuho** [comment]
@mikkelfj 
> If it's UTF-8 it is much simpler to dump in error logs, even if you don't bother validate it. Some logs tend to use JSON which is practically always UTF-8.

Don't do that. Because this is how you create a DoS attack vector on the server side.

If the reason phase contained an invalid UTF-8 sequence, creating a JSON that contains that reason phase makes the entire JSON invalid. Because many JSON decoders refuse to return anything when the input contains an invalid UTF-8 sequence, you essentially destroy all the information that were stored in JSON.

For example, consider you have an automated system that collects the reason phases for all the connections sent by the peer. Having just one invalid reason phase stops the entire system.

To summarize, calling it "MUST UTF-8" does not make things simpler.

**Comment by kazuho** [comment]
@janaiyengar 
> For devs, I can't imagine that language tags will help when the entire protocol is in English. I'm fine with a SHOULD for UTF-8, and leave the implications of non-UTF-8 encodings to implementations.

+1.

Most HTTP logs use ASCII as their output, escaping out-of-range characters using some kind of notation (e.g., `\xNN`). Assuming that the reason phrase be in English in practice, we can log the contents of the reason phase the same way as we log the HTTP headers.

**Comment by mikkelfj** [comment]
Docker tends to log to JSON. While that isn't very efficient, that is what happens. I'm not sure what kind of validation happens behind the scenes, but I'd assume it would escape anything that isn't valid UTF-8, or at least anything that contains a null character. That said, Docker is perfectly capable of DOS'ing itself without outside help.

**Comment by larseggert** [comment]
So at the moment, I think quant is the only stack that actually put something into that field. Given that other stacks aren't even doing it when we should be heavily debugging, maybe we just remove the field entirely?

**Comment by larseggert** [comment]
For debugging purposes, you could just always include a unidirectional STREAM frame with any reason text...

**Comment by mikkelfj** [comment]
I think text is probably useful for all the errors that have no precise error code as implementations become more refined and error check peer input more thoroughly. The alternative could be a request for a lot more error messages. E.g. what is protocol violation?

**Comment by dtikhonov** [comment]
> I think quant is the only stack that actually puts something into that field.

LiteSpeed QUIC puts useful information into the reason field of the CONNECTION_CLOSE frame as well.

**Comment by ianswett** [comment]
gQUIC is definitely using the reason field and intends to continue using it.

**Comment by martinduke** [comment]
@larseggert f5 sometimes puts something in there. In TCP-land we have weird hacks to include RST reasons in our packet traces, and believe this field is useful in the real world.

**Comment by MikeBishop** [comment]
@janaiyengar, the protocol isn't in English, it's in binary.  The RFC is in English, but that doesn't mean there won't be translations later.  (E.g. [RFC7540 in Chinese](https://github.com/abbshr/rfc7540-translation-zh_cn))

This is ultimately useful for the developer of the implementation emitting it, so they know what language to expect there.  The key requirement for everyone else is something easy to log and relay back to that developer, and I think UTF-8 is our best option there.  I'd support MUST UTF-8.

**Comment by janaiyengar** [comment]
(I do think the protocol is in English and Binary. The protocol is not only
what's on the wire, it's also what to do with the bits, and that's the RFC.
I'll concede that there can be translations though.)

**Comment by huitema** [comment]
I could see that some kind of text could be useful in debug more. But when running in production mode, I would just use the error code number, no text at all. If the error code is not precise enough, then we should create new error codes. The classic counter argument is that creating too many error codes provides information about the state of the stack that can be used by attackers. But then, adding explicit messages provides just as much information to attackers, so cannot possibly be better than adding more error codes...

**Comment by martinthomson** [comment]
Tokyo conclusion: We already have a SHOULD for UTF-8 text.  No language tag will be added.
